### PR TITLE
Update ENS Content

### PR DESCRIPTION
**Pull request by Kenny-Gin1**
# Description

Issue: https://github.com/DXgovernance/DAVI/issues/43
Attached: Pictures below

This PR comes with an additional core action: Updating ENS Content. Basically, the action allows the user to input his/her ENS name together with a new content hash and thus updating the content on ENS. 

E.g. `kenny1337.eth` points to the IPFS domain: `QmRAQB6YaCyidP37UdDnjFY5vQuiBrcqdyoW1CuDgwxkD4`, and we want it to point to: `QmdTPkMMBWQvL8t7yXogo7jq5pAcWg8J7RkLrDsWZHT82y`. We do so by calling the function `setContentHash` on the ENS resolver. 

- Added `UpdateENSContent`-folder with the primary components `UpdateENSContentEditor`, `UpdateENSContentInfoLine`, and `UpdateENSSummary` in line with our current ActionBuilder.
- Built a hook with the `wagmi` library to call the ENS Public Resolver contract`
- Built two new hooks: `useDebounce`to not spam an endpoint on every `onChange`-event and `useUpdateEnsContent`
- Integration tests are added, and unit tests for happy paths are added as well.
- Refactored a bit by moving styles in the `RepMint` folder to its separate file.

**TODO:**
- Add validation for the form (Separate issue)
- Make it possible to add subdomains (separate issue)
- E2E-testing
Screenshot 2022-07-26 at 15 33 08
Screenshot 2022-07-26 at 15 33 20
Screenshot 2022-07-26 at 15 33 26
Screenshot 2022-07-26 at 15 33 40

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Snapshot tests, unit tests, manual testing.

# Checklist:

- [x] My code follows the existing style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any UI changes have been tested and made responsive for mobile views

### CODE REVIEW

**Comment by Kenny-Gin1** [comment]
Design: https://www.figma.com/file/cjLng0AbOKuDj9r5oZVDmR/Wireframe?node-id=6607%3A20091

File: src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/UpdateENSContentEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+          hash: contentHash,
+        },
+      });
+    }
+  }, [debouncedIpfsHash]);
+  return (
+    <React.Fragment>
+      <Control>
+        <ControlLabel>
+          {'ENS Name'}
```
Comment (dcrescimbeni) [inline src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/UpdateENSContentEditor.tsx:69]: Add translation

File: src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/UpdateENSContentEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        </ControlLabel>
+        <ControlRow>
+          <Input value={ensName} onChange={e => setEnsName(e.target.value)} />
+          <p>.eth</p>
+        </ControlRow>
+      </Control>
+      <ControlRow>
+        <Control>
+          <ControlLabel>
+            {'CID/IPFS Hash'}
```
Comment (dcrescimbeni) [inline src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/UpdateENSContentEditor.tsx:81]: Add translation

File: src/Components/ActionsBuilder/SupportedActions/index.tsx [hunk 3]
```diff
@@ -146,6 +170,7 @@ export const getEditor = (actionType: SupportedAction) => {
 };
 
 export const displaySubmit = (actionType: SupportedAction) => {
+  console.log({ actionType });
```
Comment (dcrescimbeni) [inline src/Components/ActionsBuilder/SupportedActions/index.tsx:173]:
```suggestion
```

**Comment by dcrescimbeni** [comment]
Man, that was a mouthful of code. Awesome Kenny!

**Comment by dcrescimbeni** [comment]
The edit functionality isn't working for me

**Comment by Kenny-Gin1** [comment]
> The edit functionality isn't working for me

Hm, what exactly is not working? Is it not changing or what's happening?

**Comment by dcrescimbeni** [comment]
> > The edit functionality isn't working for me
> 
> Hm, what exactly is not working? Is it not changing or what's happening?

I can set the ENS and the content hash. Then if I go to "edit", both fields appear blank.

Here's a video

https://www.loom.com/share/24c6cc10c3c7451cb7ea639cf2a5c890

File: src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/utils.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+export const isValidChainId = (chainId: number) => {
+  if (!chainId || chainId === LOCALHOST_ID) {
+    return MAINNET_ID;
+  } else {
+    return chainId;
+  }
+};
+
+export const getBlockChainUrl = (chainId: number, address: string) => {
```
Comment (mprasanjith) [inline src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/utils.ts:34]: We now have a utility function to do this in `src/provider/chains.ts`. (`getBlockExplorerUrl`) It uses wagmi's internal config to generate the block explorer URL. Can we use that instead?

**Review by mprasanjith — CHANGES_REQUESTED** [review]
Have an interesting problem. ENS is only available on Ethereum Mainnet (excluding testnets). How are we going to handle this in DAVI?

I think we need some logic to not display the ENS update functionality in the action builder if the current chain doesn't have an ENS deployment.

File: package.json [hunk 5]
```diff
@@ -62,6 +62,7 @@
     "copy-to-clipboard": "^3.3.1",
     "dxdao-contracts": "https://github.com/DXgovernance/dxdao-contracts.git#develop",
     "eslint-plugin-cypress": "^2.12.1",
+    "eth-ens-namehash": "^2.0.8",
```
[the remark this answers is not in the document]
Comment (Kenny-Gin1) [reply package.json:65]: Definitely!

File: src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/utils.ts (hunk 4 above)
Comment (Kenny-Gin1) [reply src/Components/ActionsBuilder/SupportedActions/UpdateENSContent/utils.ts:34]: Yes, much better!

**Comment by Kenny-Gin1** [comment]
> Have an interesting problem. ENS is only available on Ethereum Mainnet (excluding testnets). How are we going to handle this in DAVI?
> 
> I think we need some logic to not display the ENS update functionality in the action builder if the current chain doesn't have an ENS deployment.

Hm, I see. I thought I read somewhere that they´re introducing it to the L2 chains. Might be in the future or something. For now we can keep it simple and only show it for mainnet + testnets.

**Comment by Kenny-Gin1** [comment]
> > > The edit functionality isn't working for me
> > 
> > 
> > Hm, what exactly is not working? Is it not changing or what's happening?
> 
> I can set the ENS and the content hash. Then if I go to "edit", both fields appear blank.
> 
> Here's a video
> 
> https://www.loom.com/share/24c6cc10c3c7451cb7ea639cf2a5c890

Thanks Dino! Will look at it!

**Comment by Kenny-Gin1** [comment]
@mprasanjith @dcrescimbeni 

Hey guys! Now I've reviewed both of your comments. They should be in the last two commits. The edit function that @dcrescimbeni pointed out was solved by adding `optionalProps` to our contract call. Since they will not be executed in the actual smart contract, it should be fine to do it this way.

If not, we could add something like `localStorage`

**Review by dcrescimbeni — APPROVED** [review]

**Pull request merged**